xbmc/addons/LanguageResource.cpp

```diff
@@ -0,0 +1,253 @@
+    if (m_dvdLanguageMenu.empty())
+      m_dvdLanguageMenu = m_language;
+    if (m_dvdLanguageAudio.empty())
+      m_dvdLanguageAudio = m_language;
+    if (m_dvdLanguageSubtitle.empty())
+      m_dvdLanguageSubtitle = m_language;
+
+    // parse <sorttokens>
+    cp_cfg_element_t *sorttokensElement = CAddonMgr::Get().GetExtElement(ext->configuration, "sorttokens");
+    if (sorttokensElement != NULL)
+    {
+      for (size_t i = 0; i < sorttokensElement->num_children; ++i)
+      {
+        cp_cfg_element_t &tokenElement = sorttokensElement->children[i];
+        if (tokenElement.name != NULL && strcmp(tokenElement.name, "token") == 0 &&
+            tokenElement.value != NULL)
+        {
+          std::string token(tokenElement.value);
+          std::string separators = CAddonMgr::Get().GetExtValue(&tokenElement, "@separators");
+          if (separators.empty())
+            separators = " ._";
+
+          for (std::string::const_iterator separator = separators.begin(); separator != separators.end(); ++separator)
+            m_sortTokens.insert(token + *separator);
+        }
+      }
+    }
+  }
+}
+
+CLanguageResource::CLanguageResource(const CLanguageResource &rhs)
+  : CResource(rhs),
+    m_locale(rhs.m_locale),
+    m_language(rhs.m_language),
+    m_territory(rhs.m_territory),
+    m_codeset(rhs.m_codeset),
+    m_modifier(rhs.m_modifier)
+{ }
+
+AddonPtr CLanguageResource::Clone() const
+{
+  return AddonPtr(new CLanguageResource(*this));
+}
+
+bool CLanguageResource::IsInUse() const
+{
+  return StringUtils::EqualsNoCase(CSettings::Get().GetString(LANGUAGE_SETTING), GetLocale());
+}
+
+bool CLanguageResource::OnPreInstall()
+{
+  return IsInUse();
+}
+
+void CLanguageResource::OnPostInstall(bool restart, bool update)
+{
+  if (restart ||
+    (!update && CGUIDialogYesNo::ShowAndGetInput(Name(), g_localizeStrings.Get(24132), "", "")))
+  {
+    CGUIDialogKaiToast *toast = (CGUIDialogKaiToast *)g_windowManager.GetWindow(WINDOW_DIALOG_KAI_TOAST);
+    if (toast)
+    {
+      toast->ResetTimer();
+      toast->Close(true);
+    }
+
+    std::string locale = GetLocaleLC();
+    if (IsInUse())
+      g_langInfo.SetLanguage(locale);
+    else
+      CSettings::Get().SetString(LANGUAGE_SETTING, locale);
+  }
+}
+
+bool CLanguageResource::IsAllowed(const std::string &file) const
+{
+  return file.empty() ||
+         StringUtils::EqualsNoCase(file.c_str(), "langinfo.xml") ||
+         StringUtils::EqualsNoCase(file.c_str(), "strings.po") ||
+         StringUtils::EqualsNoCase(file.c_str(), "strings.xml");
+}
+
+std::string CLanguageResource::GetLocaleLC() const
+{
+  std::string locale = GetLocale();
+  StringUtils::ToLower(locale);
+
+  return locale;
+}
+
+std::string CLanguageResource::GetAddonId(const std::string &locale)
+{
+  if (locale.empty())
+    return "";
+
+  std::string addonId = LANGUAGE_ADDON_PREFIX + locale;
+  StringUtils::ToLower(addonId);
+
```

Comment:
I just remembered why I chose to only store the locale in the `locale.language` setting and then based on that decided to go with a `resource://language/<locale>` path schema and therefore needed a way to create the addon ID from the locale. The value stored in `locale.language` is used in a lot of places (which I need to go through again to check their usage) and having to get the addon to retrieve the locale in every place would have been more work than just being able to use the value as the locale.